src/hotspot/share/prims/jvmtiEventController.cpp

```diff
@@ -772,7 +772,7 @@ void JvmtiEventControllerPrivate::set_event_callbacks(JvmtiEnvBase *env,
 
   env->set_event_callbacks(callbacks, size_of_callbacks);
   // Mask to clear normal event bits.
-  const jlong CLEARING_MASK = (1L >> (TOTAL_MIN_EVENT_TYPE_VAL - TOTAL_MIN_EVENT_TYPE_VAL)) - 1L;
+  const jlong CLEARING_MASK = (1L >> (JVMTI_MIN_EVENT_TYPE_VAL - TOTAL_MIN_EVENT_TYPE_VAL)) - 1L;
```

Comment:
I don't follow how this works..
Should it be "<<" instead of ">>" ?

Reply:
Maybe it would be clearer to turn off the bits when there are no events:

  jlong enabled_bits = env->env_event_enable()->_event_callback_enabled.get_bits();
  for (int ei = JVMTI_MIN_EVENT_TYPE_VAL; ei <= JVMTI_MAX_EVENT_TYPE_VAL; ++ei) {
      jvmtiEvent evt_t = (jvmtiEvent)ei;
      if (!env->has_callback(evt_t)) {
          // clear the bit if there is no callback
          enabled_bits &= ~JvmtiEventEnabled::bit_for(evt_t);
      }
  }

Reply:
Thanks, Alex! You are right - fixed.
I was stupid enough to confuse the direction. Minimal tracing helps in such cases.

Reply:
Clearing event bits to make it more readable looks like a good idea.
What about the following? :
jlong enabled_bits = env->env_event_enable()->_event_callback_enabled.get_bits();
for (int ei = JVMTI_MIN_EVENT_TYPE_VAL; ei <= JVMTI_MAX_EVENT_TYPE_VAL; ++ei) {
  jvmtiEvent evt_t = (jvmtiEvent)ei;
  if (env->has_callback(evt_t)) {
     // set the bit if there is a callback
     enabled_bits |= JvmtiEventEnabled::bit_for(evt_t);
  } else {
    // clear the bit if there is no callback
    enabled_bits &= ~JvmtiEventEnabled::bit_for(evt_t);
  }
}

Reply:
Alex, I've removed the CLEARING_MASK and implemented your suggestion which makes the code more clear.

Reply:
Thank you. It looks much clearer to me now.